go/listobjectversions.go

```diff
@@ -0,0 +1,115 @@
+package bucketclient
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"net/url"
+)
+
+type ListObjectVersionsOption func(*listObjectVersionsOptionSet) error
+
+func ListObjectVersionsMarkerOption(keyMarker string, versionIdMarker string) ListObjectVersionsOption {
+	return func(opts *listObjectVersionsOptionSet) error {
+		opts.keyMarker = keyMarker
+		opts.versionIdMarker = versionIdMarker
+		return nil
+	}
+}
+
+func ListObjectVersionsMaxKeysOption(maxKeys int) ListObjectVersionsOption {
+	return func(opts *listObjectVersionsOptionSet) error {
+		if maxKeys < 0 || maxKeys > 10000 {
+			return fmt.Errorf("maxKeys=%d is out of the valid range [0, 10000]", maxKeys)
+		}
+		opts.maxKeys = maxKeys
+		return nil
+	}
+}
+
+// ListObjectVersionsLastMarkerOption option makes the listing behave
+// as if the bucket contains no object which key/versionId is strictly
+// higher than the pair "lastKeyMarker/lastVersionIdMarker".
+//
+// Note: this option is not implemented natively by bucketd, hence the
+// Go client may truncate the result and adjust the "IsTruncated"
+// field accordingly, before returning the truncated response to the
+// client.
+func ListObjectVersionsLastMarkerOption(lastKeyMarker string, lastVersionIdMarker string) ListObjectVersionsOption {
+	return func(opts *listObjectVersionsOptionSet) error {
+		opts.lastKeyMarker = lastKeyMarker
+		opts.lastVersionIdMarker = lastVersionIdMarker
+		return nil
+	}
+}
+
+type ListObjectVersionsEntry struct {
+	Key       string `json:"key"`
+	VersionId string `json:"versionId"`
+	Value     string `json:"value"`
+}
+
+type ListObjectVersionsResponse struct {
+	Versions            []ListObjectVersionsEntry
+	CommonPrefixes      []string
+	IsTruncated         bool
+	NextKeyMarker       string `json:",omitempty"`
+	NextVersionIdMarker string `json:",omitempty"`
+}
+
+type listObjectVersionsOptionSet struct {
+	keyMarker           string
+	versionIdMarker     string
+	maxKeys             int
+	lastKeyMarker       string
+	lastVersionIdMarker string
+}
+
+func parseListObjectVersionsOptions(opts []ListObjectVersionsOption) (listObjectVersionsOptionSet, error) {
+	parsedOpts := listObjectVersionsOptionSet{
+		keyMarker:       "",
```

Comment:
In Go, fields are initiated to their zero value so this is unnecessary, but perhaps you wanted it to be extra explicit?

Reply:
Yes in this context there's no value in being explicit, I can remove them and only leave the `maxKeys` one.